default/python/AI/AIstate.py

```diff
@@ -87,6 +87,10 @@ def convert_to_version(state, version):
         for var_name in ['visInteriorSystemIDs', 'exploredSystemIDs', 'visBorderSystemIDs', 'unexploredSystemIDs']:
             state[var_name] = set(state.get(var_name, {}).keys())
 
+    elif version == 3:
```

Comment:
Make a function `state.should_surrender` at the and call it each turn.   It will not hit performance, but will make code more clear.

Reply:
I am not sure I understand what you want to say or how it is related to this particular part of the code. Maybe you can clarify what you mean?

As far as this part of the code is concerned:
The AI should keep track of the fact that it surrendered in savegames. Thus this information needs to be stored in the AIstate instance which it is with the boolean ```AIstate.surrendered```. This part of the code is supposed to provide backwards compatibility with previous savegames.

Reply:
> The AI should keep track of the fact that it surrendered in savegames. 

I suggest to calculate it each game session and don't change AIState.

Reply:
> I suggest to calculate it each game session and don't change AIState.

if you don't save it in AIstate and instead just keep it in some session-specific variable, then it will spam out its surrender messages and peace/alliance proposals anytime you load a game where one or more AIs have previously surrendered.  I don't think the player really needs a reminder of the surrender, it would just seem a little broken.  

Why do you *not* want to keep track that the AI has surrendered?

Reply:
> if you don't save it in AIstate and instead just keep it in some session-specific variable, then it will spam out its surrender messages and peace/alliance proposals anytime you load a game where one or more AIs have previously surrendered. I don't think the player really needs a reminder of the surrender, it would just seem a little broken.

Good point.  

> Why do you not want to keep track that the AI has surrendered?

I want to keep AI state as slim as is possible. 

PS. Can AI just kill all his fleets and outposts and go to terminated stage?

Reply:
> Can AI just kill all his fleets and outposts and go to terminated stage?

I don't think this is actually desirable from a game design perspective:

For fleets, this potentially robs the player of the chance to engage into that last fleet battle that they want.

For outposts, there are more severe implications for the ongoing game: The player may feel forced to conquer the outposts before the last colony in order to "optimize" because troops may be cheaper or more readily available than outposters.
It also impacts other players that may currently be in the process of invading such outposts which suddenly disappear.